### PR TITLE
mruby-regexp: apply `/i` to character classes

### PR DESCRIPTION
**Pull request by takumin**
`RE_FLAG_IGNORECASE` is folded in only where `compile_atom()` emits a single
literal character. `compile_charclass()` never reads the flag, so every `[...]`
ignores `/i`.

```ruby
/[abc]/i.match?("A")        # CRuby: true,  mruby: false
/[a-c]/i.match?("A")        # CRuby: true,  mruby: false
/[A-C]/i.match?("a")        # CRuby: true,  mruby: false
/[a-c]+/i.match?("AB")      # CRuby: true,  mruby: false
```

The inline and constructor forms of the flag reach the same code, so they fail
the same way.

```ruby
/(?i)[a-c]/.match?("A")                              # CRuby: true, mruby: false
Regexp.new("[a-c]", Regexp::IGNORECASE).match?("A")  # CRuby: true, mruby: false
```

POSIX bracket classes are affected too, since they are merged into the same
bitmap.

```ruby
/[[:lower:]]/i.match?("A")  # CRuby: true,  mruby: false
/[[:upper:]]/i.match?("a")  # CRuby: true,  mruby: false
```

A negated class is worse: it matches what it must reject, so a `/i` regexp
used for validation accepts input it should refuse. Nothing raises.

```ruby
/[^a-c]/i.match?("A")       # CRuby: false, mruby: true
```

## Fix

Fold the ASCII letters into the class bitmap at the end of
`compile_charclass()`, after the parse loop and before the class is committed
to an `RE_CLASS` or `RE_NCLASS`. That is the only point at which a `[...]` is
complete: the loop merges POSIX bracket bits, shorthands, ASCII ranges and
single literals into one bitmap, so a single pass there covers all four forms
and `[[:lower:]]` under `/i` picks up uppercase.

Negation is applied at match time against that same bitmap, so folding the
positive set fixes `[^a-c]` at the same time. `class_add_shorthand()` and
`posix_class_bits()` both run before the class is complete and would miss
plain literals and `a-c` ranges, so neither is the right place.

`class_get_bit()` is added alongside `class_set_bit()`. `re_compile.c` had no
bit-level reader for `re_charclass::bitmap`: `first_set_walk()` ORs whole
bytes, and the one bit test in the gem lives in `class_match()`, which is
`static` in another translation unit and takes a codepoint rather than a bit
index.

The fold has to happen at compile time rather than in the matcher for three
reasons. `class_match()` receives only a `re_charclass` and has no access to
any flags. `pat->flags` is the whole-pattern option set and so cannot see
inline `(?i:...)` scoping. And `compute_first_set()` derives `pat->first_bytes`
from the bitmap at compile time, which the matcher then uses to skip input; a
match time fold would leave that skip narrower than the class, and `/[a-c]/i`
would still fail on `"A"`.

Reading `c->flags` is what makes the inline forms work: `compile_atom()`
already saves, replaces and restores it around `(?i)` and `(?i:...)`, so the
class sees the scoped value with no further change.

The existing `IGNORECASE` blocks in `compile_atom()` are already correct for a
single literal and are left alone.

## Scope

Non-ASCII case folding is not addressed. The codepoint range list and
`cc->utf8_any` are untouched, and `class_add_range()` and
`class_add_codepoint()` are unchanged.

## Tests

`mrbgems/mruby-regexp/test/regexp.rb`:

- `Regexp - case insensitive character class`, a new block next to
  `Regexp - case insensitive`: the four positive forms, the
  `Regexp::IGNORECASE` constructor form, and both negated forms. It also
  guards the fold against widening the class past the ASCII letters, since
  `[` and `{` are 32 apart but are not a case pair.
- `Regexp - POSIX bracket classes`: `[[:upper:]]` and `[[:lower:]]` under
  `/i`, next to the existing `[[:upper:]]` assertion.
- `Regexp - inline options (?i) / (?i:...)`: `(?i)[a-c]`, `(?i:[a-c])`, and
  that the option does not leak past the closing paren.

Verified on `x86_64-linux`:

- Every reproduction above now matches CRuby 4.0.6.
- `rake test`: 1968 total, 1950 OK, 0 KO, 0 crash, and bintest 105 OK.
- An `MRB_INT32` build with clang and `-Wall -Wextra`: 1870 total, 1852 OK,
  0 KO, 0 crash, and no new warning from any `mruby-regexp` file.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved case-insensitive regular expression matching for ASCII letters in character classes.
  * Added support for case-insensitive matching in ranges, shorthand classes, POSIX classes, and negated classes.
  * Ensured inline case-insensitive options remain limited to their intended scope.
  * Preserved existing behavior for non-ASCII character ranges.

* **Tests**
  * Added coverage for case-insensitive character classes, negation, POSIX classes, and ASCII boundary cases.

### CODE REVIEW

File: mrbgems/mruby-regexp/src/re_compile.c [hunk 1]
```diff
@@ -466,6 +473,19 @@ compile_charclass(re_compiler *c)
...
+     applied at match time against this same bitmap (RE_NCLASS), so folding
+     the positive set also fixes [^a-c] under /i. Non-ASCII case folding is
+     out of scope, so the codepoint range list is left alone. */
+  if (c->flags & RE_FLAG_IGNORECASE) {
+    for (int ch = 'a'; ch <= 'z'; ch++) {
+      if (class_get_bit(cc, (uint8_t)ch)) class_set_bit(cc, (uint8_t)(ch - 32));
+      else if (class_get_bit(cc, (uint8_t)(ch - 32))) class_set_bit(cc, (uint8_t)ch);
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (takumin) [reply mrbgems/mruby-regexp/src/re_compile.c:488]: The behaviour is real, but the diagnosis is not: this is not a case folding gap.

`compile_charclass()` stores a range whose lower bound is ASCII and whose upper
bound is not in the codepoint list, whole. `class_match()` answers a codepoint
below 128 from the bitmap alone and never reads that list. So the ASCII half of
such a range matches nothing, with or without `/i`:

```ruby
/[a-Ā]/.match?("a")   # CRuby: true,  mruby: false
/[^a-Ā]/.match?("a")  # CRuby: false, mruby: true
```

Merging a case folded Unicode set into the class would not fix that, because
`"a"` itself is already unreachable. The fold added here is not what skips `A`
in `/[a-Ā]/i`; the range never reaches the bitmap the fold walks.

That is a separate bug, present before this PR and unchanged by it. It is fixed
in #7052, which splits a straddling range so the bitmap takes the part below
128. With both changes in, `/[a-Ā]/i` matches `"A"` and `/[^a-Ā]/i`
rejects it, agreeing with CRuby 4.0.6.

Non-ASCII case folding proper, `/[Ā]/i` against `"ā"`, stays out of
scope as this PR states. It needs Unicode case folding data, which the gem does
not carry.
